base/generator.jl

```diff
@@ -89,13 +89,19 @@ Base.HasLength()
 """
 IteratorSize(x) = IteratorSize(typeof(x))
 IteratorSize(::Type) = HasLength()  # HasLength is the default
+# Many compiled methods iterate on tuples, but MethodInstances for
+#    IteratorSize(::Type{Tuple{...}})
+# are easily invalidated by method-table insertions because they call
+# the fallback. Calling iteratorsize instead of IteratorSize bypasses
+# this problem. Don't add methods to iteratorsize.
+iteratorsize(::Type{T}) where {T} = T<:Tuple ? HasLength() : IteratorSize(T)
```

Comment:
Would it work just to add an `IteratorSize(::Type{<:Tuple})` method?

Reply:
You would know much better than me, but what I was concerned about was that it would be far down on the method list and therefore easily invalidated. But maybe order isn't as important as I think it may be?

Reply:
Perhaps because of https://github.com/JuliaLang/Pkg.jl/pull/1806 I can't see any invalidation from this anymore. I've backed this out, we can re-add it if necessary.